### ISSUE TITLE
experiment: Segmented Training to Recover MMLU

### ISSUE THREAD

**Issue by 0xSage** [homebrewltd/ichigo#59]
## Problem

From @tikikun 
After benchmarking the pretraining checkpoint on MMLU, we observed a significant degradation in the model's text capabilities. The introduction of new multilingual data caused the model's performance to drop sharply, with the accuracy falling to approximately 0.377 at the last step checkpoint --> We need to set up a series of experimental runs to attempt recovering the MMLU performance during the instruction tuning stage.

## Description

- Instruction Task (after a shorter pretrain, 3k steps)
Training time: ~ 3800 steps ~3 hours of training using 8xH100
![image](https://github.com/user-attachments/assets/576b8bdd-fe88-4827-8d39-f23e4aeae7e9)

- Instruction Task (after a longer pretrain, 9k steps)
Training time: ~ 3800 steps ~3 hours of training using 8xH100
![image](https://github.com/user-attachments/assets/01a3b84c-43ab-4e5a-bd98-d81f1b97d470)

- Instruction & Transcription Task (after a longer pretrain, 9k steps)
Training time: ~ 4800 steps ~4 hours of training using 8xH100
![image](https://github.com/user-attachments/assets/3efde3e3-48f4-4f3e-85cc-41a53822c9d6)

- It seems that further pretrain will slow down the speed of MMLU recovery, but not too far off (55 vs 51)
- Per qualitative checking, and scoring wise (alpaca/openhermes audio) it's cleary that further pretrain has positive impact on the model
ability to converge to instruction ability
- Adding filtering token and transcription in step 3 slow down the recovery speed of MMLU

## Hypothesis

![image](https://github.com/user-attachments/assets/f8943d4e-8a62-4e45-b7c4-b2cbc6bed40a)

we suspect if we teach the model to do transcription first and then do instruction later, the recovery speed will stay the same without mixing data and we’re still able to introduce 1-1 mapping in transcription.

- The model can recover drastically from MMLU 0.3 regardless of longer pretrain, we need more instruction tune data (which we already have, the test is just on subset)
- Mixed transcription may cause unwanted effects on MMLU ability -> replace transcription with segmenting strategy, the dataset will have only transcription data in the beginning and change to instruction on the latter part

**Comment by 0xSage** [comment]
Can you update results of this experiment? @tikikun @bachvudinh

**Comment by tikikun** [comment]
Model loss exploding when the data is switched from transcription -> instruction

<img width="83" alt="Screenshot 2024-09-17 at 12 36 36" src="https://github.com/user-attachments/assets/1856d034-b8e0-4ad9-a885-7433c29b9957">

Learning:
- Mixed dataset is still required
- Since the recovery is slower, we need to expand the dataset len

**Comment by tikikun** [comment]
[2e-4.txt](https://github.com/user-attachments/files/17023421/2e-4.txt)

**Comment by 0xSage** [comment]
A few more questions: 

1. Can you please provide some more details on what you observed when the model loss started exploding? What specific metrics or graphs did you use to track the issue, and what was the exact behavior you saw?

2. What are the implications now if we can't get it to work with segmented data? What are the next steps?

3. Can you elaborate on what you mean by 'recovery is slower' and how we can address it by expanding the dataset length?

I want to make sure we're making the most out of this experiment, given it was a substantial amount of burst compute. Even if things didn't work out as planned. 

The learnings will will help us learn and improve for the next iteration. Thanks!

**Comment by tikikun** [comment]
1. We segmented the dataset into two segments:
- The first 30% of the dataset contained only transcription data.
- The last 70% of the dataset contained only instruction data.

We observed that the loss stabilized during the first training run, which consisted of only the transcription data. However, when the model was converted to use instruction data, the loss spiked significantly and showed no tendency to decrease. You can find the data at [2e-4.txt](https://github.com/user-attachments/files/17023505/2e-4.txt).

2. Implications:
- We were unable to get the model to recover the MMLU score faster based on segmenting the dataset according to our theory as described in the issue description.
- A mixed dataset remains the most straightforward way for cross-domain training, at least in our case.

Next steps:
- We will revert to using a mixed dataset.
- If required, we will further explore alternative ways to segment the dataset, such as dividing it into more than two pieces. However, this would require significant additional time and computational resources, so we will prioritize reverting to the mixed dataset for now.

3. In theory, the MMLU score will recover based on the number of training steps taken linearly, as per our original observation. Therefore, if we have more data and train longer, the MMLU score can potentially recover to a much better number.

**Comment by bachvudinh** [comment]
1. When training with the data mixing strategy , we have set different learning rate from 1 to 2.5e^-4 but observed the same loss pattern that the loss seem to go down fast to 1.3 in the first 350k transcription data but when it comes to sound instruction data the loss explode and then if i keep the training continue it seem to converges to 6.9-7.0.
2. so i decided to shuffle the whole data like our previous training(v0.2) and the loss seem to go down really fast to 1.3 in just 1000 step and no sign that it will explode.

**Comment by 0xSage** [comment]
> 1. When training with the data mixing strategy , we have set different learning rate from 1 to 2.5e^-4 but observed the same loss pattern that the loss seem to go down fast to 1.3 in the first 350k transcription data but when it comes to sound instruction data the loss explode and then if i keep the training continue it seem to converges to 6.9-7.0.
> 2. so i decided to shuffle the whole data like our previous training(v0.2) and the loss seem to go down really fast to 1.3 in just 1000 step and no sign that it will explode.

This is interesting. So even after playing around with the LRs, the model couldn't converge on a new task (i.e. instruction)? 

And thank you and @hahuyhoang411 for the late-nite troubleshooting ytd.

**Comment by tikikun** [comment]
- Yes
- Yes, it is already proven required in llama3-s 0.2, without transcription task, it will confuse the tokens a bit too much and does not recognize the words that well.

**Comment by tikikun** [comment]
@bachvudinh please update final training result and we can close this ticket

**Comment by bachvudinh** [comment]
MMLU Results of latest Run that we mixed the transciprtion and instruct data equally
- Checkpoint 3000:

![Image](https://github.com/user-attachments/assets/d371687a-69e8-4cb3-82f1-90cdb1f3cf78)

- Checkpoint 4000:

![Image](https://github.com/user-attachments/assets/345b81ed-3f31-4e02-b29e-c45ce824a565)

- Checkpoint 5000:

![Image](https://github.com/user-attachments/assets/c0b76c08-90ef-42e9-8773-5156556de075)

- Checkpoint 6000:

![Image](https://github.com/user-attachments/assets/830e9887-6927-498c-9fad-efdf8aefeef0)

- Checkpoint 7000:

![Image](https://github.com/user-attachments/assets/dcb3c04a-05c3-48e4-b468-c75bde66b803)

-Checkpoint end epoch(7300):

![Image](https://github.com/user-attachments/assets/e73f8638-2302-4e4e-9b3e-59a5dbc91464)

**Comment by tikikun** [comment]
close again, as this is already properly recorded

@hahuyhoang411 for taking note in writing process.
